lib/Monads/OptionMonadWP.thy

```diff
@@ -21,8 +21,15 @@ begin
    TODO: design a sensible syntax for them. *)
 
 (* Partial correctness. *)
-definition ovalid :: "('s \<Rightarrow> bool) \<Rightarrow> ('s \<Rightarrow> 'a option) \<Rightarrow> ('a \<Rightarrow> 's \<Rightarrow> bool) \<Rightarrow> bool" where
-  "ovalid P f Q \<equiv> \<forall>s r. P s \<and> f s = Some r \<longrightarrow> Q r s"
+definition ovalid :: "('s \<Rightarrow> bool) \<Rightarrow> ('s \<Rightarrow> 'a option) \<Rightarrow> ('a \<Rightarrow> 's \<Rightarrow> bool) \<Rightarrow> bool"
+  ("o\<lbrace>_\<rbrace>/ _ /\<lbrace>_\<rbrace>") where
```

Comment:
I don't think we were happy with this syntax. I can't think of anything nicer, and I know that our options for fancy symbols are limited. Is there maybe something that looks like `{|...|}`, but replacing the curly braces with some other pair that's open and closed?

Reply:
Oh, good point. I had only been thinking getting rt in sync with master but if we can think of an alternative then this would be the right time to change that syntax.

Reply:
Right, yes, there was that. Did we come up with something in the meeting? The main problem we had was that `o{| .. |}` looks like there is a left-over function composition somewhere. The same would unfortunately apply to `{| .. |}o`. 

We could make it `⦃ .. ⦄⇩o` (subscript `o`), but that is a lot to type.

Any other ideas?

Reply:
Yeah, I thought about subscript as well but it's such a pain to type.
Looking at unused brackets, we could use `⦉P⦊ f ⦉Q⦊`. Still not the easiest to type as they don't have a unique abbreviation, but presumably that can be changed.

Reply:
The more I think about it, the less I like ovalid as the concept that it is now (nothing to do with this PR as such, it just made me think about it). The post condition should just be a about the return value, because the state can't change (by type alone)-- it doesn't make sense to carry a state predicate in the post condition and have bind rules for it etc. Of course that's now a bit late to change in AutoCorres, but maybe that is still what we should do before it gets even more ingrained.

How much ovalid are we using in the proofs?

Reply:
I'm really not sure about the use of `ovalid` in `master`, but it's used here and there for MCS. Not nearly as much as it could be used: I've had to write some pretty low level rules for `ovalid` about `get_object`, etc., but that's really just come up very recently.

I like the idea that we shouldn't mention the state in the postcondition. Before I put more rules for reader monad concepts into Lib, it might be good to change this. And I do like Corey's idea for the syntax.

Reply:
I'm not sure if I understand what the alternative is, don't we want to be able to connect return values to the state? If that's the case then how could we have the state in the pre condition but not the post? What would `obind_wp` and `ogets_wp` look like?

Separately, I think I've changed my mind. For my purposes it would be good to get something like this PR merged quickly. Are we ok with merging it as is and discussing/modifying `ovalid` in parallel?

Reply:
Yes, ok, let's not block this PR on that. Corey's syntax suggestion is a bit better, but not ideal yet, because it doesn't suggest anything about option or reader monad. If you want to get this merged quickly, I'm fine with Corey's syntax for now until we can find something better, though.

For connecting the state and the return value: you're right, you'd want the state to be available for the post condition for that purpose. My first reaction would be to just have a post condition then, but it's possible that you want to be able to split off the state-only part so you can propagate it. Maybe that's where this all came from. But do we really need to propagate it?

In either case, that might suggest a different syntax where there is no precondition, just a state + rv condition and maybe an additional state condition. We can change that after this PR, though.

One more idea for the state in this one: since the state does not change, we could just name it and leave conditions on it to normal Isabelle preconditions. E.g. something like `P s ==> f {| = s. %r. Q s r |}` where the `= s` is a free `s` and the definition of `f {| = s. Q |}` would be `!r. f s = Some r --> Q s r`. I don't know if that would work nicely, but might be worth exploring a bit.

Reply:
> I'm really not sure about the use of `ovalid` in `master`, but it's used here and there for MCS. Not nearly as much as it could be used: I've had to write some pretty low level rules for `ovalid` about `get_object`, etc., but that's really just come up very recently.

This is actually quite encouraging. If it's not used much in MCS either yet, then there is probably not much reader monad use overall. For the reader/projection proofs in master I don't actually use any special syntax at all, I just write lemmas of the form `[| f s = Some r; P s |] ==> Q s r`. Since `obind` and `opt_map` decompose nicely in that `f = Some` form, there is no further machinery. I don't know if that's ideal, it's just what I ended up doing.